### PR TITLE
Finite State Machine for Security

### PR DESCRIPTION
**Pull request by kurtuluso**
A generic FSM has been added to DDS Security Core component to realize the authentication handshake process.
The list of the states and the transitions are given in the creation and the FSM is started with a start call.
Passing arguments to transition funstions is possible.
Timeout transitions are possible.

I could not find the equivalent of global init and exit functions:
https://github.com/kurtuluso/cyclonedds/blob/fsm/src/security/core/src/fsm.c#L482
https://github.com/kurtuluso/cyclonedds/blob/fsm/src/security/core/src/fsm.c#L469

And the threads should be clean up gracefully at exit. A call back can be send to DDSI in the future:
https://github.com/kurtuluso/cyclonedds/blob/fsm/src/security/core/src/fsm.c#L487

Signed-off-by: Kurtulus Oksuztepe <kurtulus.oksuztepe@adlinktech.com>

### CODE REVIEW

File: src/security/core/src/fsm.c [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+
+    if (ut__fsm_context) {
+      /* os_procAtExit( ut__fsm_fini ); ??? */
+      ut_fsm_initialized = true;
+    } else {
+      ret = false;
+    }
+  } else {
+    if (ut_fsm_initialized == false) {
+      dds_duration_t sleep = 100000000; /* 100ms */
```
Comment (dennis-adlink) [inline src/security/core/src/fsm.c]: Use DDS_MSECS(100)

File: src/security/core/src/fsm.c [hunk 2]
```diff
@@ -0,0 +1,755 @@
...
+#include "dds/security/core/fsm.h"
+#include "dds/ddsrt/threads.h"
+#include "dds/ddsrt/sync.h"
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/process.h"
+#include "dds/ddsrt/atomics.h"
+#include "dds/ddsrt/retcode.h"
+#include "dds/ddsrt/time.h"
+#include "dds/ddsrt/timeconv.h"
+#include <assert.h>
```
Comment (dennis-adlink) [inline src/security/core/src/fsm.c]: Duplicate include

**Comment by eboasson** [comment]
Thanks @kurtuluso and @dennis-adlink for this PR and the review comments so far. I have a few of my own as well, ones that really echo those I just made for #340:
* Setup and teardown need to be invoked from ``rtps_init``/``rtps_fini`` and it would be much better to have the state in dynamically allocated memory.
* The addition of ``ddsrt_gettid_for_thread`` needs a stronger argumen than ``dds_security_fsm_free`` alone, because it is the consequence of what seems to be an overly complicated implementation. Minimising impact to the work (and work-in-progress) that depends on this FSM machinery means avoiding changes to the interface, but the way I see it, that is doable with reasonable overhead if one were to replace the list of FSMs by an index (keyed on the address) and checking whether the FSM has been freed or not after the callback got invoked. (Alternatively, one could just create a ``FREE`` state ...)
* Two threads seems like total overkill. Perhaps I overlooked something but I'd think that tracking pending timeouts is easy enough using a priority queue, which then means the ``fsm_thread`` can simply block until an event arrives or the first timeout expires. So why another thread?
* Thread creation would best be integrated more tightly with the rest of the DDSI stack code, so that thread attributes can be set like for all other threads.

File: src/security/core/src/dds_security_fsm.c [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+  for (i = 0; i < size; i++)
+  {
+    /* It needs to have a start. */
+    if (transitions[i].begin && transitions[i].event_id == DDS_SECURITY_FSM_EVENT_AUTO)
+      return true;
+  }
+  return true;
+}
+
+static void add_fsm_to_list (struct dds_security_fsm_control *control, struct dds_security_fsm *fsm)
```
Comment (eboasson) [inline src/security/core/src/dds_security_fsm.c]: I foresee trouble here, it looks like some editing mishap occurred.

File: src/security/core/src/dds_security_fsm.c [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+  ddsrt_mutex_lock (&fsm->control->lock);
+  if (fsm->active)
+  {
+    fsm_dispatch (fsm, event_id, prio);
+    ddsrt_cond_signal (&fsm->control->cond);
+  }
+  ddsrt_mutex_unlock (&fsm->control->lock);
+}
+
+const dds_security_fsm_state* dds_security_fsm_current_state (struct dds_security_fsm *fsm)
```
Comment (eboasson) [inline src/security/core/src/dds_security_fsm.c]: Is this not missing a mutex lock/unlock pair?

File: src/security/core/src/dds_security_fsm.c [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+{
+  assert(fsm);
+  assert(fsm->active);
+
+  return fsm->current;
+}
+
+void dds_security_fsm_set_debug (struct dds_security_fsm *fsm, dds_security_fsm_debug func)
```
Comment (eboasson) [inline src/security/core/src/dds_security_fsm.c]: Is this not missing a mutex lock/unlock pair?

File: src/security/core/src/dds_security_fsm.c [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+      }
+      else
+      {
+        dds_time_t now = dds_time ();
+        dur_to_wait = timeout - now;
+      }
+      if (dur_to_wait > 0)
+      {
+        thread_state_asleep (ts1);
+        (void)ddsrt_cond_waitfor( &control->cond, &control->lock, dur_to_wait);
```
Comment (eboasson) [inline src/security/core/src/dds_security_fsm.c]: There is also a ``ddsrt_cond_waituntil``

**Pull request merged**